ocaml/xapi/message_forwarding.ml

```diff
@@ -1687,9 +1687,11 @@ module Forward = functor(Local: Custom_actions.CUSTOM_ACTIONS) -> struct
             (* Intra-pool: reserve resources on the destination host, then
              * forward the call to the source. *)
             let snapshot = Db.VM.get_record ~__context ~self:vm in
-            (fun ~local_fn ~__context ~vm op ->
-               allocate_vm_to_host ~__context ~vm ~host ~snapshot ~host_op:`vm_migrate ();
-               forward_vm_op ~local_fn ~__context ~vm op)
+            fun ~local_fn ~__context ~vm op ->
+              finally
+                (fun () -> allocate_vm_to_host ~__context ~vm ~host ~snapshot ~host_op:`vm_migrate ();
+                           forward_vm_op ~local_fn ~__context ~vm op)
+                (finally_clear_host_operation ~__context ~host ~host_op:`vm_migrate);
```

Comment:
i'm not sure which of the following options is better:

```
# initial attempt
finally
  (fun () -> allocate_vm_to_host ...
                 forward_vm_op ...)
  (finally_clear_host_operation ...)
```

```
# most similar to existing code
# but if allocate_vm_to_host fails after setting things
# in the db, we will have inconsistent state
allocate_vm_to_host ...
finally
  (fun () -> foward_vm_op ...)
  (finally_clear_host_operation ...)
```

```
# what i think would be most robust
finally_with_exception
  (fun () -> allocate_vm_to_host ...
                 foward_vm_op ...)
  (fun op_exception -> match op_exception with
    | Some ex -> finally_clear_host_operation ..; raise ex
    | None -> finally_clear_host_operation )
```

Reply:
i've opened https://github.com/xapi-project/stdext/pull/40 to allow me to achieve the 3rd option above

Reply:
You could also put the finally_clear_host_operation into a try/with that just logs the exception and doesn't raise, so that your finally function never raises. That way the original exception is preserved, and we follow the semantics of the upcoming `Fun.protect` in ocaml 4.08: https://caml.inria.fr/pub/docs/manual-ocaml/libref/Fun.html

Reply:
i think we _should_ raise the exception, so as to preserve the old behaviour (unless this is definitely a bad idea?)

you're right, i shouldn't have been using finally here